File: src/lib/screening/exporters.ts
import type { ScreeningResult, ScoredCandidate } from "./types";

const CSV_COLUMNS = [
  "rank", "name", "overall_score", "match_level", "shortlist", "recommendation",
  "email", "phone", "years_experience", "education", "filename",
  "skills_30", "semantic_25", "experience_20", "education_10", "required_tech_10", "projects_5",
  "cosine_similarity", "matched_skills", "missing_skills", "missing_required_tech",
  "strengths", "gaps", "reasoning",
];

function row(c: ScoredCandidate): (string | number)[] {
  return [
    c.rank, c.profile.name, c.overall_score, c.match_level, c.shortlist ? "yes" : "no", c.recommendation,
    c.profile.email ?? "", c.profile.phone ?? "", c.profile.years_experience, c.profile.education_label, c.filename,
    c.breakdown.skills, c.breakdown.semantic, c.breakdown.experience, c.breakdown.education,
    c.breakdown.required_tech, c.breakdown.projects, c.semantic_similarity,
    c.matched_skills.join("; "), c.missing_skills.join("; "), c.missing_required_tech.join("; "),
    c.strengths.join(" | "), c.gaps.join(" | "), c.reasoning,
  ];
}

const escape = (value: string | number) => `"${String(value).replace(/"/g, '""')}"`;

export function toCsv(candidates: ScoredCandidate[]): string {
  return [CSV_COLUMNS.join(","), ...candidates.map((c) => row(c).map(escape).join(","))].join("\n");
}

export function download(filename: string, content: string, mime: string) {
  const url = URL.createObjectURL(new Blob([content], { type: mime }));
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  anchor.click();
  URL.revokeObjectURL(url);
}

export function downloadCsv(result: ScreeningResult) {
  download(`screening-${result.run_id.slice(0, 8)}.csv`, toCsv(result.candidates), "text/csv");
}

export function downloadJson(result: ScreeningResult) {
  download(
    `screening-${result.run_id.slice(0, 8)}.json`,
    JSON.stringify(result, null, 2),
    "application/json",
  );
}

File: backend/exporters.py
"""CSV / JSON export helpers for a completed screening run."""

from __future__ import annotations

import csv
import io
import json

CSV_COLUMNS = [
    "rank", "name", "overall_score", "match_level", "shortlist", "recommendation",
    "email", "phone", "years_experience", "education", "filename",
    "skills_30", "semantic_25", "experience_20", "education_10", "required_tech_10", "projects_5",
    "cosine_similarity", "matched_skills", "missing_skills", "missing_required_tech",
    "strengths", "gaps", "reasoning",
]


def results_to_rows(candidates) -> list[dict]:
    rows = []
    for c in candidates:
        rows.append({
            "rank": c.rank,
            "name": c.profile.name,
            "overall_score": c.overall_score,
            "match_level": c.match_level,
            "shortlist": "yes" if c.shortlist else "no",
            "recommendation": c.recommendation,
            "email": c.profile.email or "",
            "phone": c.profile.phone or "",
            "years_experience": c.profile.years_experience,
            "education": c.profile.education_label,
            "filename": c.filename,
            "skills_30": c.breakdown.skills,
            "semantic_25": c.breakdown.semantic,
            "experience_20": c.breakdown.experience,
            "education_10": c.breakdown.education,
            "required_tech_10": c.breakdown.required_tech,
            "projects_5": c.breakdown.projects,
            "cosine_similarity": c.semantic_similarity,
            "matched_skills": "; ".join(c.matched_skills),
            "missing_skills": "; ".join(c.missing_skills),
            "missing_required_tech": "; ".join(c.missing_required_tech),
            "strengths": " | ".join(c.strengths),
            "gaps": " | ".join(c.gaps),
            "reasoning": c.reasoning,
        })
    return rows


def to_csv(candidates) -> str:
    buffer = io.StringIO()
    writer = csv.DictWriter(buffer, fieldnames=CSV_COLUMNS)
    writer.writeheader()
    writer.writerows(results_to_rows(candidates))
    return buffer.getvalue()


def to_json(payload: dict) -> str:
    return json.dumps(payload, indent=2, default=str)
